### PR TITLE
Support Custom CA for certificate issuing

### PR DESCRIPTION
**Pull request by meng-han**
previously we have one giant class certificatemanager which has a certificatauthority class which is basically ACM PCA.

Now we want to add support for custom ca, without impacting existing contracts/apis on `/v1/certificates prefix`. So a new base class is created so that both acm pca and custom ca can inherit from.

other changes are mostly tests and settings along side this.

### CODE REVIEW

**Review by skiptomyliu — APPROVED** [review]

File: confidant/services/certificate_authority/customca.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+                )
+            )
+        return active_ca[0]
+
+    def _load_ca_certificate(self, ca_json):
+        return x509.load_pem_x509_certificate(ca_json["crt"].encode("utf-8"))
+
+    def _load_rootca_certificate(self, ca_json):
+        if "rootcrt" not in ca_json or not ca_json["rootcrt"]:
+            logger.warning("Custom CA has no root CA certificate provided")
```
Comment (Yueren-Wang) [inline confidant/services/certificate_authority/customca.py:94]: maybe adding ca id to enrich the log for debugging purpose?
logger.warning("Custom CA %s has no root CA certificate provided", ca_env)

File: confidant/services/certificate_authority/customca.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        builder = builder.issuer_name(
+            self.ca_certificate.subject
+        )  # Issued by our CA
+        builder = builder.public_key(csr.public_key())
+        builder = builder.serial_number(x509.random_serial_number())
+        builder = builder.not_valid_before(datetime.now(timezone.utc))
+
+        acceptable_validity = min(validity, self.settings["max_validity_days"])
+        builder = builder.not_valid_after(
+            datetime.now(timezone.utc) + timedelta(days=acceptable_validity)
```
Comment (Yueren-Wang) [inline confidant/services/certificate_authority/customca.py:133]: call datetime.now twice at 129 and 133. considering consolidate them into 1?

File: confidant/services/certificate_authority/customca.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+            raise CertificateAuthorityNotFoundError(
+                f"Custom CA {ca_env} has no active keys"
+            )
+        validator = Validator(CUSTOM_CA_SCHEMA)
+        active_ca_id = CUSTOM_CA_ACTIVE_KEYS[ca_env]
+        self.active_ca_id = active_ca_id
+        active_ca = [
+            ca
+            for ca in CUSTOM_CERTIFICATE_AUTHORITIES[ca_env]
+            if validator.validate(ca) and ca["kid"] == active_ca_id
```
Comment (Yueren-Wang) [inline confidant/services/certificate_authority/customca.py:77]: if validator.validate(ca) return False, will we still log Custom CA %s has no active keys?

File: confidant/services/certificate_authority/customca.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+            return None
+        return x509.load_pem_x509_certificate(
+            ca_json["rootcrt"].encode("utf-8")
+        )
+
+    def _load_ca_chain(self):
+        # Get the certificate in PEM format
+        intermediate_ca_pem = self.encode_certificate(self.ca_certificate)
+        root_ca_pem = self.encode_certificate(self.root_ca_certificate)
+        return intermediate_ca_pem + root_ca_pem
```
Comment (Yueren-Wang) [inline confidant/services/certificate_authority/customca.py:104]: self.root_ca_certificate can be None? if so intermediate_ca_pem + root_ca_pem will throw TypeError

**Review by Yueren-Wang — APPROVED** [review]

**Review by Yueren-Wang — APPROVED** [review]

**Pull request merged**